src/workerd/api/encoding.c++

```diff
@@ -464,25 +465,95 @@ jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
-  };
-}
-}  // namespace
-
 jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+  jsg::JsString str = input.orDefault(js.str());
+
+  if (str.isOneByte(js)) {
+    // Fast path for one-byte strings (Latin-1). writeOneByte() copies the raw bytes without
+    // flattening the string, which is more efficient than using ValueView. Note that we
+    // allocate `length * 2` bytes because Latin-1 characters 0x80-0xFF need 2 bytes in UTF-8.
+    auto backing = jsg::BackingStore::alloc<v8::Uint8Array>(
+        js, str.length(js) * 2, jsg::Lock::AllocOption::UNINITIALIZED);
+    str.writeOneByte(js, backing.asArrayPtr<kj::byte>());
+    return jsg::BufferSource(js, kj::mv(backing));
+  }
+
+  // First pass: Calculate the required UTF-8 output buffer size.
+  // We need to do this in a separate ValueView because:
+  // 1. ValueView holds the V8 heap lock, which prevents us from allocating new V8 objects
+  // 2. We must determine the exact output size before allocating the BackingStore
+  // 3. Once we know the size, we'll create a second ValueView to do the actual conversion
+  size_t utf8_length = 0;
+  bool isValidUtf16 = true;
+  // For invalid UTF-16 strings (with unpaired surrogates), we need to fix them to well-formed
+  // UTF-16 before calculating the UTF-8 length. We store the fixed version here so it can be
+  // reused in the second pass, avoiding the need to fix it twice.
+  kj::Array<char16_t> wellFormed;
+
+  {
+    v8::String::ValueView view(js.v8Isolate, str);
+    // One-byte strings are handled by the fast path above
+    KJ_DASSERT(!view.is_one_byte());
+
+    auto data = reinterpret_cast<const char16_t*>(view.data16());
+    // Two-byte string path. V8 uses UTF-16LE encoding internally for strings with code points
+    // > U+00FF. Check if the UTF-16 is valid (no unpaired surrogates) to determine the path.
+    isValidUtf16 = simdutf::validate_utf16le(data, view.length());
+
+    if (isValidUtf16) {
+      // Common case: valid UTF-16, calculate UTF-8 length directly
+      utf8_length = simdutf::utf8_length_from_utf16le(data, view.length());
+    } else {
+      // Rare case: Invalid UTF-16 with unpaired surrogates. Per the Encoding Standard,
+      // unpaired surrogates must be replaced with U+FFFD (replacement character).
+      // U+FFFD is 3 bytes in UTF-8, which means the UTF-8 length will differ from what
+      // we'd calculate from the invalid UTF-16. We must fix the UTF-16 first, then
+      // calculate the UTF-8 length from the well-formed version to get the correct size.
+      wellFormed = kj::heapArray<char16_t>(view.length());
```

Comment:
Note that there's a risk here. If the original string wasn't flat, creatig the `ValueView` will flatten it, doubling the heap usage. Then, you're allocating yet another copy here... then you allocate the target backing store below, which at worst case is by itself potentially a 3x increase of the original allocation (worst case Nx6)